<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [82]:
import pandas as pd
import psycopg2

In [83]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [84]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [85]:
# текст запроса
query_3_1 = f'''SELECT 
                    COUNT(id) vacancies
                FROM 
                    vacancies
             '''

In [86]:
# результат запроса
quantity_vacancies = pd.read_sql_query(query_3_1, connection)
quantity_vacancies

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\505335110.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  quantity_vacancies = pd.read_sql_query(query_3_1, connection)


vacancies
0      49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [87]:
# текст запроса
query_3_2 = f'''SELECT 
                    COUNT(id) employers
                FROM 
                    employers
             '''

In [88]:
# результат запроса
quantity_employers = pd.read_sql_query(query_3_2, connection)
quantity_employers

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\72524357.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  quantity_employers = pd.read_sql_query(query_3_2, connection)


employers
0      23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [89]:
# текст запроса
query_3_3 = f'''SELECT 
                    COUNT(id) areas
                FROM 
                    areas
             '''

In [90]:
# результат запроса
quantity_areas = pd.read_sql_query(query_3_3, connection)
quantity_areas

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\2486803467.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  quantity_areas = pd.read_sql_query(query_3_3, connection)


areas
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [92]:
# текст запроса
query_3_4 = f'''SELECT 
                    COUNT(id) industres
                FROM 
                    industries
             '''

In [93]:
# результат запроса
quantity_industres = pd.read_sql_query(query_3_4, connection)
quantity_industres

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\2956326683.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  quantity_industres = pd.read_sql_query(query_3_4, connection)


industres
0        294

***

In [ ]:
# выводы по предварительному анализу данных

- среднее количество вакансий у работодателя с вакансиями: 3.3 (общее количество вакансий 49197 и количество работодателей с вакансиями 14906). А вообще среднее число вакансий на всех работодателей: 2.1.
- 25 компаний имеют количество вакансий от 100 и более, это всего 0.1 % от общего количества работодателей с вакансиями.
- максимальное количество вакансий составляет 1933 у компании Яндекс.
- распределение работодателей по регионам - прямая корреляция с количеством населения, административным статусом, количеством экономических субъектов. В первую сотню по количеству вакансий входят административные центры субъектов Российской Федерации, остальные крупнейшие города, а также столицы и крупные города стран СНГ.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [94]:
# текст 
query_4_1 = f'''SELECT 
                a.name,
                COUNT(v.id) cnt
            FROM vacancies v JOIN areas a ON v.area_id = a.id
            GROUP BY a.name
            ORDER BY COUNT(v.id) DESC
            LIMIT 5
            '''


In [95]:
# результат запроса
quantity_vacancies_5 = pd.read_sql_query(query_4_1, connection)
quantity_vacancies_5

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\3541048108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  quantity_vacancies_5 = pd.read_sql_query(query_4_1, connection)


name   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [96]:
# текст запроса
query_4_2 = f'''SELECT
                    COUNT(id)
                FROM
                    vacancies 
                WHERE
                    salary_from IS NOT NULL OR salary_to IS NOT NULL
            '''

In [97]:
# результат запроса
quantity_vacancies = pd.read_sql_query(query_4_2, connection)
quantity_vacancies

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\1719400490.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  quantity_vacancies = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [98]:
# текст запроса
query_4_3 = f'''SELECT
                    ROUND(AVG(salary_from))::int avg_salary_from,
                    ROUND(AVG(salary_to))::int avg_salary_to
                FROM
                    vacancies
             '''

In [99]:
# результат запроса
avg_salary = pd.read_sql_query(query_4_3, connection)
avg_salary

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\212845365.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0            71065         110537

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [100]:
# текст запроса
query_4_4 = f'''WITH sched_employment AS (
                    SELECT
                        schedule, 
                        employment,                     
                        COUNT(id) vacancies
                    FROM
                        vacancies 
                    GROUP BY schedule, employment
                )
                SELECT
                    schedule, 
                    employment,                     
                    vacancies,
                    ROUND(vacancies/(SELECT SUM(vacancies) FROM sched_employment)*100, 5) percent
                FROM
                    sched_employment
                ORDER BY vacancies DESC
             '''

In [101]:
# результат запроса
schedule_employment = pd.read_sql_query(query_4_4, connection)
schedule_employment

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\1022405654.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  schedule_employment = pd.read_sql_query(query_4_4, connection)


schedule           employment  vacancies   percent
0        Полный день     Полная занятость      35367  71.88853
1   Удаленная работа     Полная занятость       7802  15.85869
2      Гибкий график     Полная занятость       1593   3.23800
3   Удаленная работа  Частичная занятость       1312   2.66683
4     Сменный график     Полная занятость        940   1.91069
5        Полный день           Стажировка        569   1.15657
6     Вахтовый метод     Полная занятость        367   0.74598
7        Полный день  Частичная занятость        347   0.70533
8      Гибкий график  Частичная занятость        312   0.63419
9        Полный день     Проектная работа        141   0.28660
10  Удаленная работа     Проектная работа        133   0.27034
11     Гибкий график           Стажировка        116   0.23579
12    Сменный график  Частичная занятость        101   0.20530
13  Удаленная работа           Стажировка         64   0.13009
14     Гибкий график     Проектная работа         18   0.03659
15    Сменный график           Стажировка         12   0.02439
16    Вахтовый метод     Проектная работа          2   0.00407
17    Сменный график     Проектная работа          1   0.00203

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [103]:
# текст запроса
query_4_5 = f'''WITH exp_how_many AS (
                    SELECT
                        experience,
                        COUNT(id) quantity
                    FROM
                        vacancies
                    GROUP BY experience
                )
                SELECT
                    experience,
                    quantity,
                    ROUND(quantity/(SELECT SUM(quantity) FROM exp_how_many)*100, 1) percent
                FROM
                    exp_how_many
                ORDER BY quantity
            '''

In [104]:
# результат запроса
experience_vacancies = pd.read_sql_query(query_4_5, connection)
experience_vacancies

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\4185389635.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  experience_vacancies = pd.read_sql_query(query_4_5, connection)


experience  quantity  percent
0         Более 6 лет      1337      2.7
1           Нет опыта      7197     14.6
2       От 3 до 6 лет     14511     29.5
3  От 1 года до 3 лет     26152     53.2

***

In [ ]:
# выводы по детальному анализу вакансий

- Распределение вакансий по регионам прямо коррелирует с количеством населения, административным статусом, количеством экономических субъектов в данном регионе. В первую сотню по количеству вакансий входят административные центры субъектов Российской Федерации, крупнейшие города кроме них, а также столицы и крупные города стран СНГ. Так на долю Москвы приходится почти 11 % от общего количества вакансий, доля Санкт-Петербурга почти 6 %. На третьем месте Минск с более, чем 4 %.
- Более половины вакансий (51 % - все-таки более) не имеет каких-либо указаний на предлагаемую заработную плату, не минимума ни максимума.
- Средние значения для нижней и верхней границы зарплатной вилки составляют 71 и 110 тыс. рублей.
- Более 80 % вакансий предполагают личное присутствие на рабочем месте (полный день, сменный и гибкий график), на долю удаленной работы приходится лишь около 19 %. Вахтовый метод где-то 0.8 %. Это странная тенденция для IT-специальностей, где многие работают удаленно.
- Требуемый опыт работы указан в подавляющем количестве вакансий - более 85 %.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [105]:
# текст запроса
query_5_1 = f'''WITH employers AS
                    SELECT
                        e.name,
                        COUNT(v.id) quantity
                    FROM
                        vacancies v
                        LEFT JOIN employers e ON v.employer_id = e.id
                    GROUP BY e.id
                    ORDER BY quantity DESC
                )
                (SELECT
                     name,
                     quantity,
                     1 place
                 FROM
                     employers
                 LIMIT 1)
                UNION
                (SELECT
                     name,
                     quantity,
                     5
                 FROM
                     employers
                 OFFSET 4
                 LIMIT 1)
                 ORDER BY place
            '''

In [106]:
# результат запроса
employer_vacancies = pd.read_sql_query(query_5_1, connection)
employer_vacancies

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\101337004.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employer_vacancies = pd.read_sql_query(query_5_1, connection)


name  quantity  place
0         Яндекс      1933      1
1  Газпром нефть       331      5

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [108]:
# текст запроса
query_5_2 = f'''SELECT
                    a.name,
                    vacan.vacan_quantity,
                    employ.employ_quantity
                FROM areas a LEFT JOIN (SELECT 
                                    a.id id,
                                    COUNT(v.id) vacan_quantity
                                FROM
                                    vacancies v 
                                    LEFT JOIN areas a ON v.area_id = a.id
                                GROUP BY a.id
                                ) vacan ON a.id = vacan.id
                     LEFT JOIN (SELECT
                                    a.id id,
                                    COUNT(e.id) employ_quantity
                                FROM
                                    employers e 
                                    LEFT JOIN areas a ON e.area = a.id
                                GROUP BY a.id
                                ) employ ON a.id = employ.id
                WHERE employ_quantity IS NOT NULL
                ORDER BY vacan_quantity DESC, employ_quantity DESC
             '''
# результат запроса

In [110]:
# результат запроса
area_vacan = pd.read_sql_query(query_5_2, connection)
area_vacan

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\3694018842.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  area_vacan = pd.read_sql_query(query_5_2, connection)


name  vacan_quantity  employ_quantity
0                Россия             NaN              410
1             Казахстан             NaN              207
2    Московская область             NaN               75
3    Краснодарский край             NaN               19
4    Ростовская область             NaN               18
..                  ...             ...              ...
649         Электроугли             1.0                1
650         Белореченск             1.0                1
651       Красный Сулин             1.0                1
652            Заволжье             1.0                1
653       Ликино-Дулево             1.0                1

[654 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [111]:
# текст запроса
query_5_3 = f'''SELECT
                    e.name,
                    COUNT(DISTINCT v.area_id) quantity
                FROM vacancies v LEFT JOIN employers e ON v.employer_id = e.id
                GROUP BY e.id
                ORDER BY quantity DESC
            '''

In [112]:
# результат запроса
employ_area = pd.read_sql_query(query_5_3, connection)
employ_area

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\1777136720.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employ_area = pd.read_sql_query(query_5_3, connection)


name  quantity
0                                      Яндекс       181
1                                  Ростелеком       152
2                                  Спецремонт       116
3                      Поляков Денис Иванович        88
4                                   ООО ЕФИН         71
...                                       ...       ...
14901                            НПП Авиатрон         1
14902              Центр дистанционных торгов         1
14903  Городские Телекоммуникационные Системы         1
14904                       Введенский, Отель         1
14905              Группа Компаний Автоимпорт         1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [121]:
# текст запроса
query_5_4 = f'''SELECT
                    COUNT(*)
                FROM employers e FULL JOIN employers_industries e_i ON e.id = e_i.employer_id
                WHERE e_i.industry_id IS NULL
            '''

In [122]:
# результат запроса
employ_empty = pd.read_sql_query(query_5_4, connection)
employ_empty

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\191027772.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employ_empty = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [123]:
# текст запроса
query_5_5 = f'''SELECT
                    e.name
                FROM employers_industries e_i LEFT JOIN employers e ON e_i.employer_id = e.id
                GROUP BY e.id
                HAVING COUNT(e_i.industry_id) = 4
                ORDER BY name
                OFFSET 2
                LIMIT 1
            '''

In [124]:
# результат запроса
third_employ = pd.read_sql_query(query_5_5, connection)
third_employ

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\89550183.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  third_employ = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [126]:
# текст запроса
query_5_6 = f'''SELECT
                    COUNT(DISTINCT employer_id)
                FROM employers_industries e_i LEFT JOIN industries i ON e_i.industry_id = i.id
                WHERE i.name = 'Разработка программного обеспечения'
             '''

In [127]:
# результат запроса
employ_industry = pd.read_sql_query(query_5_6, connection)
employ_industry

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\588175935.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  employ_industry = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [139]:
# код получения списка городов-милионников
city_list = ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', 'Пермь', 'Волгоград')

In [141]:
# текст запроса
query_5_6 = f'''WITH tab AS (
                SELECT
                    a.name,
                    COUNT(*)::int cnt
                FROM vacancies v LEFT JOIN employers e ON v.employer_id = e.id
                    LEFT JOIN areas a ON v.area_id = a.id
                WHERE e.name LIKE 'Яндекс' AND a.name IN {city_list}
                GROUP BY a.id)
                SELECT*
                FROM tab
                UNION
                SELECT 'Total',
                SUM(cnt)
                FROM tab
                ORDER BY cnt
             '''


In [142]:
# результат запроса
yandex_areas = pd.read_sql_query(query_5_6, connection)
yandex_areas

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\2561736876.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  yandex_areas = pd.read_sql_query(query_5_6, connection)


name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

In [ ]:
# выводы по анализу работодателей


- Работодатели на первом и пятом месте по количеству вакансий: Яндекс и Газпромнефть
- 186 регионов без вакансий.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [150]:
# текст запроса
# текст запроса
query_6_1 = f'''SELECT
                    COUNT(*)
                FROM vacancies v
                WHERE lower(v.name) ~* 'data' OR lower(v.name) ~* 'данн'
            '''

In [151]:
# результат запроса
data = pd.read_sql_query(query_6_1, connection)
data

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\1969111781.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [152]:
# текст запроса
query_6_2 = f'''SELECT
                    COUNT(*)
                FROM vacancies
                WHERE (name ~* 'data scientist' OR name ~* 'data science' OR 
                name ~* 'исследователь данных' OR (name ~ 'ML' AND name !~* 'HTML') OR
                name ~* 'machine learning' OR name ~* 'машинн.*обучен.*')
                AND (lower(name) ~* 'junior' OR experience ='Нет опыта' OR
                employment = 'Стажировка')
             '''   

In [153]:
# результат запроса
junior_vacancies = pd.read_sql_query(query_6_2, connection)
junior_vacancies

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\625073081.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  junior_vacancies = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [154]:
# текст запроса
query_6_3 = f'''SELECT
                    COUNT(*)
                FROM vacancies
                WHERE (name ~* 'data scientist' OR name ~* 'data science' OR 
                name ~* 'исследователь данных' OR (name ~ 'ML' AND name !~* 'HTML') OR
                name ~* 'machine learning' OR name ~* 'машинн.*обучен.*') AND
                (key_skills ~* 'SQL' OR key_skills ~* 'postgres')
             '''

In [156]:
# результат запроса
sql = pd.read_sql_query(query_6_3, connection)
sql

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\1864025986.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  sql = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [157]:
# текст запроса
query_6_4 = f'''SELECT
                    COUNT(*)
                FROM vacancies
                WHERE (name ~* 'data scientist' OR name ~* 'data science' OR 
                name ~* 'исследователь данных' OR (name ~ 'ML' AND name !~* 'HTML') OR
                name ~* 'machine learning' OR name ~* 'машинн.*обучен.*') AND (key_skills ~ 'Python')
             '''

In [158]:
# результат запроса
python = pd.read_sql_query(query_6_4, connection)
python

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\3988841599.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  python = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [159]:
# текст запроса
query_6_5 = f'''SELECT
                ROUND(AVG(length(key_skills) - length(replace(key_skills, '\t', '')) + 1), 2) skills
                FROM vacancies
                WHERE name ~* 'data scientist' OR name ~* 'data science' OR 
                name ~* 'исследователь данных' OR (name ~ 'ML' AND name !~* 'HTML') OR
                name ~* 'machine learning' OR name ~* 'машинн.*обучен.*'
             '''

In [160]:
# результат запроса
avg_skills = pd.read_sql_query(query_6_5, connection)
avg_skills

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\4266198567.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_skills = pd.read_sql_query(query_6_5, connection)


skills
0    6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [161]:
# текст 
query_6_6 = f'''SELECT
                    experience,
                    ROUND(AVG((COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)) / 2))::int avg_salary -- слагаемые дадут нам или две минимальных или две максимальных, ну или минимальную и максимальную зарплаты, да на 2 поделим, вот и будет или средняя или просто минимум или максимум
                FROM vacancies
                WHERE (name ~* 'data scientist' OR name ~* 'data science' OR 
                name ~* 'исследователь данных' OR (name ~ 'ML' AND name !~* 'HTML') OR
                name ~* 'machine learning' OR name ~* 'машинн.*обучен.*')
                AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
                GROUP BY experience
                ORDER BY avg_salary DESC
             '''

In [162]:
# результат запроса
avg_salary = pd.read_sql_query(query_6_6, connection)
avg_salary

C:\Users\pavel\AppData\Local\Temp\ipykernel_11388\4147299713.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  avg_salary = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0       От 3 до 6 лет      243115
1  От 1 года до 3 лет      139675
2           Нет опыта       74643

***

In [ ]:
# выводы по предметному анализу

- Мы имеем прямую корреляцию заработной платы с опытом работы, чем больше опыт - тем больше заработная плата. Сотрудникам без опыта готовы платить около 75 000 рублей. С опытом от года и до 3 лет - почти в два раза больше, 140 000 руб. А от 3 и до 6 лет: примерно в три с половиной раза больше, 243 000 руб. 
Для опыта более 6 лет, зарплата не указана. Очевидно, что она ещё выше.
Стоит отметить, что из 480 вакансий для дата-сайентистов, заработную плату указали лишь в 62 вакансиях.

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

- Общее количество вакансий 49 197, общее количество работодателей 23 501, а с вакансиями только 14 906. Т.е. среднее количество вакансий у работодателя: 3.3
- Распределение вакансий по регионам, коррелирует с экономикой региона. В первую сотню по количеству входят административные центры субъектов Российской Федерации, крупнейшие города страны, а также столицы и крупные города стран СНГ. На Москву приходится почти 11 % от общего количества вакансий, доля Санкт-Петербурга почти 6 %. На третьем месте Минск с более, чем 4 %.
- Более половины вакансий - 51 %, не имеют четких значение заработной платы. В 21 % имеется полная информация о предложении работодателя по зарплате. В отношении только минимума определились в 23 % вакансий, а максимума в 5 % вакансий. Средние значения для нижней и верхней границы зарплаты составляют 71 и 110 тыс. рублей.
- Более 80 % вакансий предполагают присутствие на рабочем месте (полный день, сменный и гибкий график), на долю удаленной работы приходится около 19 %. Вахтовый метод примерно 0.8 %.
- Требуемый опыт работы указан в подавляющем количестве вакансий - более 85 %. Половина устанавливает опыт от 1 до 3 лет, почти 30 % от 3 до 6 лет и 2.7 % - больше 6 лет опыта. Опыт прямо связан зарплатными предложениями, при росте опыта растет предлагаемая зарплата.
- Около 8.5 тыс. работодателей не указали сферы деятельности, в которых они предлагают работу.
- Отсутствует связь между количеством ключевых навыков и зарплатой.

Специальность «Data Science» пока ещё редко востребована из общего количества числа вакансий ИТ сферы. Вакансии предлагаются крупными компаниями. Востребованы специалисты с опытом, с набором определенных ключевых навыков. Вакансии предлагаются в крупных городах, Москве, СПб. Возможная удаленная работа. Заработная плата в целом в два – три раза выше, чем в среднем для ИТ сферы. И зависит в первую очередь от опыта.
